packages/ui/src/components/popover/popover.tsx

```diff
@@ -0,0 +1,212 @@
+  triggerMode: PopoverTriggerMode.Click,
+  className: 'popover',
+  closeOnEsc: true,
+  closeOnClick: true,
+};
+
+interface PopoverProps {
+  onClose?: () => void;
+  onOpen?: () => void;
+  closeOnClick: boolean;
+  closeOnEsc: boolean;
+  verticalPosition: PopoverVerticalPositions;
+  horizontalPosition: PopoverHorizontalPositions;
+  className: string;
+  triggerMode: PopoverTriggerMode;
+  content: JSX.Element;
+  children: JSX.Element | JSX.Element[];
+}
+
+interface PopoverState {
+  open: boolean;
+}
+
+/**
+ * Popover renders popup with a target.
+ * Properties description:
+ *   onOpen - function that will be executed when popover opens;
+ *   onClose - function that will be executed when popover closes;
+ *   closeOnClick - close the popover on mouse click;
+ *   verticalPosition - vertical position of popover relative to target ("top" or "bottom"), default - "top";
+ *   horizontalPosition - horizontal position of popover relative to target ("left" or "right"), default - "left";
+ *   className - popover class, can be use for popover styling, default - 'popover';
+ *   closeOnEsc - handle ESC keyboard pressed event and close popover, default - true;
+ *   triggerMode - popover support 2 ways for opening: click (keyPress) and mouse hover ("click" or "hover"), default - "click";
+ *   content - content of popover. Can be any valid Preact node, ex:
+ *      <select>
+ *          <option>Option 1</option>
+ *          ...
+ *          <option>Option n</option>
+ *      </select>
+ *   children - popover target. Can be any valid Preact node, ex:
+ *      <button>
+ *          <i className="icon" />
+ *      </button>
+ */
+
+export class Popover extends Component<PopoverProps, PopoverState> {
+  private _closeTimeout: any = null;
+  private _controlElement = null;
+  static defaultProps = {
+    ...defaultProps,
+  };
+  state = {
+    open: false,
+  };
+
+  private _clearTimeout = () => {
+    clearTimeout(this._closeTimeout);
+    this._closeTimeout = null;
+  };
+
+  private _handleMouseEvent = (event: MouseEvent) => {
+    if (
+      !this._controlElement.contains(event.target) &&
+      this.props.closeOnClick
+    ) {
+      this._closePopover();
+    }
+  };
+
+  private _handleKeyboardEvent = (event: KeyboardEvent) => {
+    if (
+      this._controlElement.contains(event.target) &&
+      event.keyCode === KeyboardKeys.Enter
+    ) {
+      // handle Enter key pressed on Target icon to prevent triggering of _closePopover twice
+      return;
+    }
+    if (
+      (this.props.closeOnEsc && event.keyCode === KeyboardKeys.Esc) ||
+      event.keyCode === KeyboardKeys.Enter
+    ) {
+      // handle if ESC or Enter button presesd
+      this._closePopover();
+    }
+  };
+
+  private _openPopover = () => {
+    const {onOpen} = this.props;
+    this._clearTimeout();
+    this.setState({open: true}, () => {
+      if (onOpen) {
+        onOpen();
+      }
+      this._addListeners();
```

Comment:
consider adding listeners first so if the user throw unexpected error in `onOpen` it will not prevent the registeration